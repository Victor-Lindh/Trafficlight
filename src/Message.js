import React from 'react'

function Message(props) {

    return (
        <div>
            <p id="msgDiv" className={props.time < 10 || props.time > 19 ? "doNotCross" : "pleaseCross"}>{props.note}</p>
        </div>
    )};

    export default Message;