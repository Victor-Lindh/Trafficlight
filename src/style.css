body {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  background: -webkit-gradient(linear, left top, left bottom, from(black), to(darkblue));
  background: linear-gradient(to bottom, black, darkblue);
  height: 100vh;
}

.container {
  margin: auto;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  border: 5px solid darkslategrey;
  border-radius: 5px;
  width: 200px;
  height: 85vh;
  padding: 5px 10px;
  background: -webkit-gradient(linear, left bottom, left top, from(black), to(#282828));
  background: linear-gradient(to top, black, #282828);
}

.light {
  width: 150px;
  height: 150px;
  border-radius: 50%;
  margin: 5px 10px;
  border: 2px solid black;
  border-top: 10px solid #181818;
  padding-top: 2px;
}

.red-light {
  background: -webkit-gradient(linear, left top, left bottom, from(red), color-stop(red), to(salmon));
  background: linear-gradient(to bottom, red, red, salmon);
}

.yellow-light {
  background: -webkit-gradient(linear, left top, left bottom, from(yellow), color-stop(yellow), to(lightyellow));
  background: linear-gradient(to bottom, yellow, yellow, lightyellow);
}

.green-light {
  background: -webkit-gradient(linear, left top, left bottom, from(green), color-stop(green), to(lightgreen));
  background: linear-gradient(to bottom, green, green, lightgreen);
}

.highlight-red {
  -webkit-box-shadow: 0 0 70px red;
          box-shadow: 0 0 70px red;
}

.highlight-yellow {
  -webkit-box-shadow: 0 0 70px yellow;
          box-shadow: 0 0 70px yellow;
}

.highlight-green {
  -webkit-box-shadow: 0 0 70px #5df15d;
          box-shadow: 0 0 70px #5df15d;
}

button {
  border-radius: 2px;
  outline: 1px solid darkslategray;
  margin-top: 10px;
  background-color: gray;
  color: snow;
  font-weight: bold;
  padding: 5px;
}

button[disabled] {
  background-color: black;
  color: rgba(255, 255, 255, 0.5);
}

.timer {
  background-color: black;
  color: snow;
  font-size: 1.5em;
  width: 100%;
  text-align: center;
  font-weight: bolder;
}

.message, .doNotCross, .pleaseCross {
  text-align: center;
  font-weight: bold;
  font-size: 1.5em;
  margin-top: 1px;
  margin-bottom: 2px;
  border-radius: 2px;
  border: 1px solid white;
}

.doNotCross {
  color: snow;
  background-color: firebrick;
}

.pleaseCross {
  color: snow;
  background-color: forestgreen;
}
/*# sourceMappingURL=style.css.map */