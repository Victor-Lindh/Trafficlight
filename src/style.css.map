{
    "version": 3,
    "mappings": "AAAA,AAAA,IAAI,CAAA;EACA,OAAO,EAAE,IAAI;EACb,cAAc,EAAE,MAAM;EACtB,WAAW,EAAE,MAAM;EACnB,UAAU,EAAE,2CAA4C;EACxD,MAAM,EAAE,KAAK;CAChB;;AACD,AAAA,UAAU,CAAA;EACN,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,cAAc,EAAE,MAAM;EACtB,WAAW,EAAE,MAAM;EACnB,MAAM,EAAE,uBAAuB;EAC/B,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,QAAQ;EACjB,UAAU,EAAE,uCAAuC;CACtD;;AACD,AAAA,MAAM,CAAA;EACF,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,KAAK;EACb,aAAa,EAAE,GAAG;EAClB,MAAM,EAAE,QAAQ;EAChB,MAAM,EAAE,eAAe;EACvB,UAAU,EAAE,kBAAkB;EAC9B,WAAW,EAAE,GAAG;CACnB;;AACD,AAAA,UAAU,CAAA;EACN,UAAU,EAAE,4CAA4C;CAC3D;;AACD,AAAA,aAAa,CAAA;EACT,UAAU,EAAE,uDAAuD;CACtE;;AACD,AAAA,YAAY,CAAA;EACR,UAAU,EAAE,oDAAoD;CACnE;;AACD,AAAA,cAAc,CAAA;EACV,UAAU,EAAE,YAAY;CAC3B;;AACD,AAAA,iBAAiB,CAAA;EACb,UAAU,EAAE,eAAe;CAC9B;;AACD,AAAA,gBAAgB,CAAA;EACZ,UAAU,EAAE,CAAC,CAAC,CAAC,CAAC,IAAI,CAAC,OAAgB;CACxC;;AACD,AAAA,MAAM,CAAA;EACF,aAAa,EAAE,GAAG;EAClB,OAAO,EAAE,uBAAuB;EAChC,UAAU,EAAE,IAAI;EAChB,gBAAgB,EAAE,IAAI;EACtB,KAAK,EAAE,IAAI;EACX,WAAW,EAAE,IAAI;EACjB,OAAO,EAAE,GAAG;CACf;;AACD,AAAA,MAAM,CAAA,AAAA,QAAC,AAAA,EAAS;EACZ,gBAAgB,EAAE,KAAK;EACvB,KAAK,EAAE,wBAAoB;CAC9B;;AACD,AAAA,MAAM,CAAA;EACF,gBAAgB,EAAE,KAAK;EACvB,KAAK,EAAE,IAAI;EACX,SAAS,EAAE,KAAK;EAChB,KAAK,EAAE,IAAI;EACX,UAAU,EAAE,MAAM;EAClB,WAAW,EAAE,MAAM;CACtB;;AACD,AAAA,QAAQ,EASR,WAAW,EAKX,YAAY,CAdJ;EACJ,UAAU,EAAE,MAAM;EAClB,WAAW,EAAE,IAAI;EACjB,SAAS,EAAE,KAAK;EAChB,UAAU,EAAE,GAAG;EACf,aAAa,EAAE,GAAG;EAClB,aAAa,EAAE,GAAG;EAClB,MAAM,EAAE,eAAe;CAC1B;;AACD,AAAA,WAAW,CAAA;EAEP,KAAK,EAAE,IAAI;EACX,gBAAgB,EAAE,SAAS;CAC9B;;AACD,AAAA,YAAY,CAAA;EAER,KAAK,EAAE,IAAI;EACX,gBAAgB,EAAE,WAAW;CAChC",
    "sources": [
        "style.scss"
    ],
    "names": [],
    "file": "style.css"
}