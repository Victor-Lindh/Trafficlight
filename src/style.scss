body{
    display: flex;
    flex-direction: column;
    align-items: center;
    background: linear-gradient(to bottom, black, darkblue );
    height: 100vh;
}
.container{
    margin: auto;
    display: flex;
    flex-direction: column;
    align-items: center;
    border: 5px solid darkslategrey;
    border-radius: 5px;
    width: 200px;
    height: 85vh;
    padding: 5px 10px;
    background: linear-gradient(to top, black, #282828);
}
.light{
    width: 150px;
    height: 150px;
    border-radius: 50%;
    margin: 5px 10px;
    border: 2px solid black;
    border-top: 10px solid #181818;
    padding-top: 2px;
}
.red-light{
    background: linear-gradient(to bottom, red, red, salmon);
}
.yellow-light{
    background: linear-gradient(to bottom, yellow, yellow, lightyellow);
}
.green-light{
    background: linear-gradient(to bottom, green, green, lightgreen);
}
.highlight-red{
    box-shadow: 0 0 70px red;
}
.highlight-yellow{
    box-shadow: 0 0 70px yellow;
}
.highlight-green{
    box-shadow: 0 0 70px rgb(93, 241, 93);
}
button{
    border-radius: 2px;
    outline: 1px solid darkslategray;
    margin-top: 10px;
    background-color: gray;
    color: snow;
    font-weight: bold;
    padding: 5px;
}
button[disabled]{
    background-color: black;
    color: rgba(255,255,255,.5);
}
.timer{
    background-color: black;
    color: snow;
    font-size: 1.5em;
    width: 100%;
    text-align: center;
    font-weight: bolder;
}
.message{
    text-align: center;
    font-weight: bold;
    font-size: 1.5em;
    margin-top: 1px;
    margin-bottom: 2px;
    border-radius: 2px;
    border: 1px solid white;
}
.doNotCross{
    @extend .message;
    color: snow;
    background-color: firebrick;
}
.pleaseCross{
    @extend .message;
    color: snow;
    background-color: forestgreen; 
}