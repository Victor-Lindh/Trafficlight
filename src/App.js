import React, {Component} from 'react';
import Message from './Message';
import '../src/style.scss';

class TrafficLight extends Component {
  constructor(props){
    super(props);
    this.state = {
      timer: 0,
      clicked: false,
      messageIsVisible: false,
      cars: []
    }

    this.crossTheStreet = this.crossTheStreet.bind(this);
    this.carsPopulating = this.carsPopulating.bind(this);

  }
  
    crossTheStreet(){
        this.setState({
          clicked: true,
          messageIsVisible: true
        })
        
        let time = setInterval(() => {
          this.setState({
            timer: this.state.timer + 1,
        });
        }, 1000);
        setTimeout(() => {
          clearInterval(time);
          this.setState({
            timer: 0,
            clicked: false,
            cars: []
          });
        }, 20000);
      }

 UNSAFE_componentWillUpdate(){
   this.carsPopulating();

}
    async carsPopulating () {

        let call = await this.state.timer === 3;
        if(call){
        
         this.setState({cars: ['Car 1', 'Car 2', 'Car 3']});
        }
      }
  render(){
    return (
    <>
      <div className="container">
        <div className={(this.state.timer <= 7 || this.state.timer > 19) ? "light red-light highlight-red" : "light red-light"}></div>
        <div className={(this.state.timer >= 6 && this.state.timer < 10) || (this.state.timer >= 15 && this.state.timer <= 19) ? "light yellow-light highlight-yellow" : "light yellow-light"}></div>
        <div className={(this.state.timer >= 10 && this.state.timer <= 17) ? "light green-light highlight-green" : "light green-light"}></div>
        <button id="btn" disabled={this.state.clicked} onClick={this.crossTheStreet}>Press to cross the street</button>
        <div className="timer">{this.state.timer}</div>
      </div>
        <Message
          time={this.state.timer}
          clicked={this.state.clicked}
          note={(this.state.timer < 10 || this.state.timer > 19) ? "Please do not cross!" : "Please cross!"}
        />
        <div
          style={{
            listStyleType: `none`,
            position: `absolute`, 
            top: `50px`, 
            right: `200px`, 
            color: `yellow`, 
            fontSize: `24px`,
            
            }}>
          <div>{
              (this.state.cars.length > 0 && this.state.timer >= 9 || this.state.timer >= 19) ? this.state.cars.map((car, index) => 
              <li key={index}>{car} stopped...</li>)
              : 'Cars driving...'
          }</div>
        </div>
    </>
  );
}}
export default TrafficLight;