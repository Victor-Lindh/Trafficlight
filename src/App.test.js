import React from 'react';
import { shallow, mount, render } from 'enzyme';
import TrafficLight from '../src/App';

import {crossTheStreet, carsPopulating, clearCars} from '../src/App';

beforeEach(() => {
    jest.resetModules();
    });
    
    describe('App rendering elements', () => {
        it('Render 8 divs', () => {
            const App = mount(<TrafficLight />);
            expect(App.find('div')).toHaveLength(8);
        });
        it('Render one button', () => {
            const App = render(<TrafficLight />);
            expect(App.find('button')).toHaveLength(1);
        });
    });

    describe('Test state properties for correct initial values', () => {
        it('"clicked" state property should be false', () => {
            const App = shallow(<TrafficLight />);
            expect(App.state('clicked')).toBeFalsy();
        });
        it('"timer" state property should be set to 0', () => {
            const App = shallow(<TrafficLight />);
            expect(App.state('timer')).toBe(0);
        });
        it('"cars" state property should be empty array', () => {
            const App = shallow(<TrafficLight />);
            expect(App.state('cars')).toEqual([]);
        })
        
    });

    describe('Test for state changes', () => {
        test('button click should set state property "clicked" to "true"', () => {
            const App = mount(<TrafficLight />);
            expect(App.state('clicked')).toBe(false);
            App.find('button').simulate('click', {
                target: {clicked: true}
            });
            expect(App.state('clicked')).toBe(true);
        });

        test('state variable "timer" should increment by 1 every second', () => {
            const App = shallow(<TrafficLight />);
            jest.useFakeTimers();
            expect(App.state("timer")).toEqual(0);
            App.instance().crossTheStreet();
            jest.advanceTimersByTime(3000);  // Testing out to see if counter increments by 3 seconds (arbitrarily)
            expect(App.state("timer")).toEqual(3);
            jest.useRealTimers();
        });

        test('"timer" state variable should revert to 0 after timer ends', () => {
            const App = mount(<TrafficLight />);
            jest.useFakeTimers();
            expect(App.state('timer')).toEqual(0);
            App.instance().crossTheStreet();
            jest.advanceTimersByTime(20000);
            expect(App.state('timer')).toEqual(0);
            jest.useRealTimers();
        });
        test('"clicked" state variable should revert to "false" after timer ends', () => {
            const App = mount(<TrafficLight />);
            expect(App.state('clicked')).toBe(false);
            jest.useFakeTimers();
            App.find('button').simulate('click', {
                target: {clicked: true}
            });
            expect(App.state('clicked')).toBe(true);
            jest.advanceTimersByTime(20000);
            expect(App.state('clicked')).toBe(false);
            jest.useRealTimers();
        });
    });

    describe('Snapshot testing', () => {
        test('should match snapshot', () => {
            const App = shallow(<TrafficLight />);
            expect(App).toMatchSnapshot();
        });
    });

    test('button should recieve "disabled" attribute when clicked', () => {
        const App = mount(<TrafficLight />);
        App.find('#btn').simulate('click');
        expect(App.find('#btn').get(0).props.disabled).toBe(true);
    })
    