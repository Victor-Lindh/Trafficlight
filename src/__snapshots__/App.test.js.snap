// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing should match snapshot 1`] = `
<Fragment>
  <div
    className="container"
  >
    <div
      className="light red-light highlight-red"
    />
    <div
      className="light yellow-light"
    />
    <div
      className="light green-light"
    />
    <button
      disabled={false}
      id="btn"
      onClick={[Function]}
    >
      Press to cross the street
    </button>
    <div
      className="timer"
    >
      0
    </div>
  </div>
  <Message
    clicked={false}
    note="Please do not cross!"
    time={0}
  />
  <div
    style={
      Object {
        "color": "yellow",
        "fontSize": "24px",
        "listStyleType": "none",
        "position": "absolute",
        "right": "200px",
        "top": "50px",
      }
    }
  >
    <div>
      Cars driving...
    </div>
  </div>
</Fragment>
`;
